src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs

```diff
@@ -0,0 +1,1845 @@
+            CompileAndVerify(comp, expectedOutput:
+@"System.Action<System.Object>
+System.Func<System.Object>");
+        }
+
+        /// <summary>
+        /// Custom modifiers should not affect delegate signature.
+        /// </summary>
+        [Fact]
+        public void CustomModifiers_02()
+        {
+            var sourceA =
+@".class public A
+{
+  .method public static void F1(object modreq(int32) x) { ldnull throw }
+  .method public static object modreq(int32) F2() { ldnull throw }
+}";
+            var refA = CompileIL(sourceA);
+
+            var sourceB =
+@"using System;
+class B
+{
+    static void Report(Delegate d)
+    {
+        Console.WriteLine(d.GetDelegateTypeName());
+    }
+    static void Main()
+    {
+        Report(A.F1);
+        Report(A.F2);
+    }
+}";
+            var comp = CreateCompilation(new[] { sourceB, s_utils }, new[] { refA }, parseOptions: TestOptions.RegularPreview, options: TestOptions.ReleaseExe);
+            comp.VerifyDiagnostics(
+                // (10,16): error CS1503: Argument 1: cannot convert from 'method group' to 'Delegate'
+                //         Report(A.F1);
+                Diagnostic(ErrorCode.ERR_BadArgType, "A.F1").WithArguments("1", "method group", "System.Delegate").WithLocation(10, 16),
+                // (11,16): error CS1503: Argument 1: cannot convert from 'method group' to 'Delegate'
+                //         Report(A.F2);
+                Diagnostic(ErrorCode.ERR_BadArgType, "A.F2").WithArguments("1", "method group", "System.Delegate").WithLocation(11, 16));
+        }
+
+        [Fact]
+        public void SystemActionAndFunc_Missing()
+        {
+            var sourceA =
+@".assembly mscorlib
+{
+  .ver 0:0:0:0
+}
+.class public System.Object
+{
+  .method public hidebysig specialname rtspecialname instance void .ctor() cil managed { ret }
+}
+.class public abstract System.ValueType extends System.Object
+{
+  .method public hidebysig specialname rtspecialname instance void .ctor() cil managed { ret }
+}
+.class public System.String extends System.Object
+{
+  .method public hidebysig specialname rtspecialname instance void .ctor() cil managed { ret }
+}
+.class public sealed System.Void extends System.ValueType
+{
+  .method public hidebysig specialname rtspecialname instance void .ctor() cil managed { ret }
+}
+.class public sealed System.Boolean extends System.ValueType
+{
+  .method public hidebysig specialname rtspecialname instance void .ctor() cil managed { ret }
+}
+.class public sealed System.Int32 extends System.ValueType
+{
+  .method public hidebysig specialname rtspecialname instance void .ctor() cil managed { ret }
+}
+.class public abstract System.Delegate extends System.Object
+{
+  .method public hidebysig specialname rtspecialname instance void .ctor() cil managed { ret }
+}";
+            var refA = CompileIL(sourceA, prependDefaultHeader: false, autoInherit: false);
```

Comment:
nit: Is it necessary to use IL to test those types are missing? (also in test below) 
It would be good  to also test `Expression e = () => {};` with missing `Expression`1`

Reply:
The IL is simple enough for these missing types so that seemed easiest.

Added a test for `Expression<TDelegate>`.